proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,717 @@
+  uint32_t ShaderRegister;
+  uint32_t RegisterSpace = 0;
+  hlsl::RootSignature::ShaderVisibility ShaderVisibility =
+      hlsl::RootSignature::ShaderVisibility::All;
+};
+
+struct ParsedRootSignature {
+  RootSignature::RootSignatureVersion Version;
+  uint32_t RootFlags;
+  llvm::SmallVector<RootSignature::ContainerRootParameter, 8>
+      RSParameters;
+  llvm::SmallVector<RootSignature::StaticSamplerDesc, 8> StaticSamplers;
+};
+
+}
+
+```
+
+This will require more work for print and serialization the AST node.
+
+The AST node serialization code might be looked like this:
+```
+void ASTDeclWriter::VisitHLSLRootSignatureDecl(HLSLRootSignatureDecl *D) {
+  VisitNamedDecl(D);
+  Record.AddSourceLocation(D->getKeywordLoc());
+  hlsl::ParsedRootSignature &RS = D->getRootSignature();
+  Record.push_back(RS.RSDesc.Version);
+  Record.push_back(RS.RSDesc.Flags);
+  for (const auto &P : ParsedRS.RSParameters) {
+    // push_back all the fields of root parameters.
+  }
+  for (const auto &SS : ParsedRS.StaticSamplers) {
+    // push_back all the fields of static samplers.
+  }
+  for (const auto &Range : ParsedRS.DescriptorRanges) {
+    // push_back all the descriptor ranges.
+  }
+  
+  Code = serialization::DECL_HLSL_ROOT_SIGNATURE;
+}
+
+```
+
+
+A HLSLRootSignatureAttr will be created when meet RootSignature attribute in
+HLSL.
+
+```
+
+[RootSignature(MyRS1)]
+float4 main(float4 coord : COORD) : SV_Target
+{
+…
+}
+
+```
+
+Because the RootSignature attribute in hlsl only have the string, it is easier
+to add a StringArgument to save the string first.
+A HLSLRootSignatureDecl will be created and added to the HLSLRootSignatureAttr 
+for diagnostic and saved to the HLSLEntryRootSignatureAttr for clang 
+code generation.
+The HLSLRootSignatureDecl will save StringLiteral instead StringRef for 
+diagnostic.
+
+In clang code generation, the HLSLRootSignatureAttr in AST will be translated
+into metadata to express the layout and things like static sampler, root 
+flags, space and NumDescriptors in LLVM IR.
+
+CGHLSLRuntime will generate metadata to link the metadata as root
+signature for given entry function. 
+
+For case compile to a standalone root signature blob with option like:
+```
+
+dxc.exe /T rootsig_1_1 MyRS1.hlsl /E RS /Fo MyRS1.fxo
+```
+ the HLSLRootSignatureAttr will be bind to a fake empty entry.
+
+
+When In the case of local or global root signatures like:
+```
+
+GlobalRootSignature MyGlobalRootSignature =
+{
+    "DescriptorTable(UAV(u0)),"                     // Output texture
+    "SRV(t0),"                                      // Acceleration structure
+    "CBV(b0),"                                      // Scene constants
+    "DescriptorTable(SRV(t1, numDescriptors = 2))"  // Static index and vertex buffers.
+};
+
+LocalRootSignature MyLocalRootSignature = 
+{
+    "RootConstants(num32BitConstants = 4, b1)"  // Cube constants 
+};
+
+```
+ only a HLSLRootSignatureDecl will be created.
```

Comment:
How are global/local root signatures represented in the AST?

Reply:
As a HLSLRootSignatureDecl  node.